R/downloadbuttons.R

```diff
@@ -0,0 +1,541 @@
+## MetaInsight - downloadable files for data uploading instructions - code.
+
+
+## Sizing functions for forest plots ##
+BayesPixels <- function(notrt, title=FALSE) {    # input is total number of treatments and whether title is included in plot
+  if (title==FALSE) {
+    if (notrt <= 25) {
+      height <- 420        # default for 25 or less treatments
+    } else {
+      height <- 15*(notrt-1) + 60
+    }
+  } else {
+    if (notrt <= 25) {
+      height <- 420 + 100        # default for 25 or less treatments
+    } else {
+      height <- 15*(notrt-1) + 60 + 100
+    }
+  }
+  return(height)
+}
+BayesInch <- function(notrt) {
+  if (notrt <= 25) {
+    height <- 6
+  } else {
+    height <- 6 + 0.2*(notrt-25)
+  }
+  return(height)
+}
+
+write_to_pdf_or_png <- function(file, type, renderFunction) {
+  if (tolower(type) == "pdf") {
+    pdf(file = file)
+  } else {
+    png(file = file)
+  }
+  renderFunction()
+  dev.off()
+}
+
+
+#########################
+### Tab 2 - Load data ###
+#########################
+
+create_raw_data_download_handlers <- function(input, output) {
+  
+  create_raw_data_download_handler <- function(filename, continuousFile, bindaryFile) {
+    return(
+      downloadHandler(
+        filename = filename,
+        content = function(file) {
+          if (input$metaoutcome=='Continuous') {
+            file.copy(continuousFile, file)
+          } else {
+            file.copy(bindaryFile, file)
+          }
+        }
+      )
+    )
+  }
+  
+  ##### in the 'upload long data' tab
+  output$downloadData <- create_raw_data_download_handler("MetaInsightdataLONG.csv", "Cont_long.csv", "Binary_long.csv")
```

Comment:
I feel like it isn't a good idea to have shiny outputs in this file (even if they work) as it won't be obvious where to find them.

Reply:
I ideally `output` and `input` should only be moved outside of the `server.R` and `ui.R`, if using shiny modules, for testing and to prevent non reactive values being passed to the functions.

Reply:
This file already had all of the outputs in it, nothing new has been added. This might just be my inexperience with Shiny, but I don't see a reason why outputs can't be outside of `server.R`. Every other coding project I have worked on has used good package structuring to break up logic into manageable chunks. Am I missing something obvious that makes Shiny different?

Long term, I'd like to see this all go into modules, but that is a bigger task than what was meant to be a small refactor when I started. Are we happy for this to remain as-is as an intermediate step?

Reply:
I would prefer all the `output` calls to go back into `server.R` for the same as reason @RyanJField.